File: solutions/c/say/1/say.c
#include "say.h"

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>

const char* large[] = {"billion", "million", "thousand"};
const char* small[] = {"twenty", "thirty",  "forty",  "fifty",
                       "sixty",  "seventy", "eighty", "ninety"};
const char* tiny[] = {"one",     "two",       "three",    "four",     "five",
                      "six",     "seven",     "eight",    "nine",     "ten",
                      "eleven",  "twelve",    "thirteen", "fourteen", "fifteen",
                      "sixteen", "seventeen", "eighteen", "nineteen"};

int say(int64_t input, char** ans) {
    if (input < MIN_SAY || input > MAX_SAY) return -1;

    *ans = calloc(MAX_SAY_LENGTH, sizeof(char));
    int64_t divisor = 1000000000;
    int large_idx = 0;
    int written = 0;

    while (divisor >= 1) {
        int group = input / divisor;

        if (group > 0) {
            int hundreds = group / 100;

            if (hundreds > 0) written += sprintf(*ans + written, "%s hundred ", tiny[hundreds - 1]);

            group -= hundreds * 100;
            int tens = group / 10;

            if (tens == 1) {
                written += sprintf(*ans + written, "%s ", tiny[group - 1]);
            } else {
                if (tens > 0)  written += sprintf(*ans + written, "%s", small[tens - 2]);

                int ones = group - tens * 10;
                if (ones > 0) written += sprintf(*ans + written, "%s%s ", tens > 0 ? "-" : "", tiny[ones - 1]);
            }

            if (large_idx < 3) written += sprintf(*ans + written, "%s ", large[large_idx]);
        }

        input %= divisor;
        divisor /= 1000;
        large_idx++;
    }

    if ((*ans)[written - 1] == ' ') (*ans)[written - 1] = '\0';
    if (written == 0) sprintf(*ans, "zero");

    return 0;
}

File: solutions/python/say/1/say.py
MAX_SAY = 999999999999
MIN_SAY = 0

PLACES = ["billion", "million", "thousand"]
TENS = ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
ONE_N_TEENS = [
    "",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "ten",
    "eleven",
    "twelve",
    "thirteen",
    "fourteen",
    "fifteen",
    "sixteen",
    "seventeen",
    "eighteen",
    "nineteen",
]


def say(number):
    if not (MIN_SAY <= number <= MAX_SAY):
        raise ValueError("input out of range")
    if number == 0:
        return "zero"

    divisor = 1000000000
    places_idx = 0
    says = []

    while number >= 1:
        to_say = number // divisor

        if to_say > 0:
            if to_say // 100 > 0:
                says.append(f"{ONE_N_TEENS[to_say // 100]} hundred")
                to_say %= 100

            if to_say // 10 > 1:
                ones = f"-{ONE_N_TEENS[to_say % 10]}" if to_say % 10 > 0 else ""
                says.append(f"{TENS[to_say // 10 - 2]}{ones}")
                to_say = 0
            elif to_say // 10 > 0:
                says.append(f"{ONE_N_TEENS[to_say]}")
                to_say = 0

            if to_say > 0:
                says.append(f"{ONE_N_TEENS[to_say]}")

            if places_idx < len(PLACES):
                says.append(f"{PLACES[places_idx]}")

        number %= divisor
        divisor //= 1000
        places_idx += 1

    return " ".join(says)
